Source/Core/VideoBackends/Vulkan/CommandBufferManager.cpp

```diff
@@ -160,28 +132,76 @@ void CommandBufferManager::DestroyCommandBuffers()
       vkDestroyFence(device, resources.fence, nullptr);
   }
 
-  for (VkDescriptorPool descriptor_pool : m_descriptor_pools)
+  for (FrameResources& resources : m_frame_resources)
   {
-    if (descriptor_pool != VK_NULL_HANDLE)
+    for (VkDescriptorPool descriptor_pool : resources.descriptor_pools)
+    {
       vkDestroyDescriptorPool(device, descriptor_pool, nullptr);
+    }
   }
 
   vkDestroySemaphore(device, m_present_semaphore, nullptr);
 }
 
 VkDescriptorSet CommandBufferManager::AllocateDescriptorSet(VkDescriptorSetLayout set_layout)
 {
-  VkDescriptorSetAllocateInfo allocate_info = {VK_STRUCTURE_TYPE_DESCRIPTOR_SET_ALLOCATE_INFO,
-                                               nullptr, GetCurrentDescriptorPool(), 1, &set_layout};
+  VkDescriptorSet descriptor_set = VK_NULL_HANDLE;
+  FrameResources& resources = GetCurrentFrameResources();
 
-  VkDescriptorSet descriptor_set;
-  VkResult res =
-      vkAllocateDescriptorSets(g_vulkan_context->GetDevice(), &allocate_info, &descriptor_set);
-  if (res != VK_SUCCESS)
+  if (!resources.descriptor_pools.empty()) [[likely]]
+  {
+    VkDescriptorSetAllocateInfo allocate_info = {
+        VK_STRUCTURE_TYPE_DESCRIPTOR_SET_ALLOCATE_INFO, nullptr,
+        resources.descriptor_pools[resources.current_descriptor_pool_index], 1, &set_layout};
+
+    VkResult res =
+        vkAllocateDescriptorSets(g_vulkan_context->GetDevice(), &allocate_info, &descriptor_set);
+    if (res != VK_SUCCESS &&
+        resources.descriptor_pools.size() > resources.current_descriptor_pool_index + 1)
+    {
+      // Mark the next descriptor set as active and try again.
+      resources.current_descriptor_pool_index++;
+      descriptor_set = AllocateDescriptorSet(set_layout);
+    }
+  }
+
+  if (descriptor_set == VK_NULL_HANDLE) [[unlikely]]
   {
-    // Failing to allocate a descriptor set is not a fatal error, we can
-    // recover by moving to the next command buffer.
-    return VK_NULL_HANDLE;
+    const u32 SETS_PER_POOL = 256;
```

Comment:
Would it make sense to have this grow?  e.g. `256 << resources.descriptor_pools.size()`

Reply:
That would make us create progressively larger descriptor pools. I don't really see why we'd want to do that.

Keep in mind that we don't free the old pools when creating a new one.

FWIW we could instead simply raise this number. I haven't really done a ton of testing, I just picked it and ran with it. IIRC Mario Galaxy ended up needing 2-3 pools.

Reply:
> That would make us create progressively larger descriptor pools. I don't really see why we'd want to do that.

Just to make sure we don't end up needing tons of pools, since games can be pretty variable with how many sets they need.  If a game happens to do something that requires e.g. 10000 descriptor sets, the current code would end up with 40 pools, vs 7 if you made them progressively larger.

Reply:
What we could do instead is check whether there's more than 1 descriptor pool when resetting them and if that's the case, destroy it + create a new larger one.

That way it would amortize to just one descriptor pool.

Reply:
I tested a couple of games and most seem to come out at <1024 descriptor sets.
Skyward Sword had the most sets at 2048.

I'll change it to "merge" the pools when we begin a new frame and bump the default set size.

Reply:
Sounds good

BTW if you want a game that uses a lot of pools, Densha de GO! Shinkansen EX needed 20 pools in the old code, putting it at around 5000

Reply:
Yeah, 256 sets per pool was just too low in general.